### PR TITLE
(EAI-353): Quiz question evaluation

### PR DESCRIPTION
**Pull request by mongodben**
Jira: https://jira.mongodb.org/browse/EAI-353

## Changes

In `mongodb-chatbot-evaluation` package:

- Quiz question data generator (i.e. get response from model)
- Quiz question evaluator (i.e. is model response correct)
- Have pipeline functions return metadata with possibility of additional info.
- Add createdAt field to `GeneratedData`

New `model-eval` package for running evals on LLMs including:

- Create package w/ package config (based on `chatbot-eval-mongodb-public`)
- Config to run model quiz question evaluations.
- Support calling multiple models thru Radiant. 
  - Currently just Azure OpenAI models. Hopefully will have more soon...

In `chatbot-eval-mongodb-public` package:

- Support new pipeline function format.

## Notes

-

### CODE REVIEW

File: packages/mongodb-chatbot-evaluation/package.json [hunk 1]
```diff
@@ -42,7 +42,8 @@
     "lint": "eslint ./src --ext ts --report-unused-disable-directives",
     "lint:fix": "npm run lint -- --fix && prettier ./src --check --write",
     "release": "release-it",
-    "test": "jest --forceExit"
+    "test": "jest --forceExit",
+    "shuffle": "node ./build/generate/shuffleQuizData.js"
```
Comment (nlarew) [inline packages/mongodb-chatbot-evaluation/package.json:46]: What's this script? I don't see the corresponding file anywhere

Comment (mongodben) [reply packages/mongodb-chatbot-evaluation/package.json:46]: oh we can remove that, i made that as a 1-off when creating quiz question dataset

Comment (mongodben) [inline packages/mongodb-chatbot-evaluation/package.json:46]:
```suggestion
    "test": "jest --forceExit"
```

Comment (mongodben) [reply packages/mongodb-chatbot-evaluation/package.json]: removed. closing comment

File: packages/model-eval/src/radiantModels.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  {
+    label: "gpt-4o",
+    radiantModelDeployment: "gpt-4o-eai-experimentation",
+  },
+  // TODO: mistral seems broken
+  // {
+  //   label: "mistral-large",
+  //   radiantModelDeployment: "Mistral-large-eai",
+  // },
```
Comment (nlarew) [inline packages/model-eval/src/radiantModels.ts:25]: EAI for this?

File: packages/mongodb-chatbot-evaluation/src/test/mockGenerateDataFunc.ts [hunk 3]
```diff
@@ -1,40 +1,68 @@
...
+  const generatedData: ConversationGeneratedData[] =
+    // Note: not sure why I need to cast this to ConversationTestCase[]
+    // The code still compiles in editor without this,
+    // but it doesn't at run-time in the tests.
+    // Very strange, but I don't think it's problematic.
+    (
+      testCases.filter((testCase) =>
+        isConversationTestCase(testCase)
+      ) as ConversationTestCase[]
+    )
```
Comment (nlarew) [inline packages/mongodb-chatbot-evaluation/src/test/mockGenerateDataFunc.ts:30]:
```suggestion
    (
      testCases.filter((testCase): testCase is ConversationTestCase =>
        isConversationTestCase(testCase)
      )
    )
```

I think this is likely the issue/solution - the `.filter()` function doesn't narrow the return type so you have to define the type predicate yourself with the `is` operator.

The next version of TS (v5.5) adds support for [Inferred Type Predicates](https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#inferred-type-predicates) which will do this automatically for us once we upgrade.

File: packages/mongodb-chatbot-evaluation/src/generate/generateLlmQuizQuestionAnswer.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  sleepMs?: number;
+}
+
+/**
+  Generate answer to multiple-choice quiz questions using a large language model (LLM).
+
+  This can be useful for evaluating how an LLM performance on the subject matter of the multiple choice questions.
+
+  The prompt is based on this blog post from Hugging Face: https://huggingface.co/blog/open-llm-leaderboard-mmlu
+  It follows the [HELM prompt format](https://huggingface.co/blog/open-llm-leaderboard-mmlu#mmlu-comes-in-all-shapes-and-sizes-looking-at-the-prompts).
```
Comment (nlarew) [inline packages/mongodb-chatbot-evaluation/src/generate/generateLlmQuizQuestionAnswer.ts:61]: Would it be worth splitting this out into a reusable `makeHelmPrompt()` function w/ accompanying tests? Not necessary for this PR but might be useful down the line + as documentation of the format.

**Review by nlarew — APPROVED** [review]
LGTM w/ a few non-blocking comments

File: packages/model-eval/.env.example [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+MONGODB_DATABASE_NAME="..."
+# Note: The quiz questions are stored in staging
+MONGODB_QUIZ_QUESTIONS_CONNECTION_URI="mongodb+srv://<user>:<pass>@docs-chatbot-staging.ngvo1.mongodb.net/?retryWrites=true&w=majority"
+MONGODB_QUIZ_QUESTIONS_DATABASE_NAME="datasets"
+MONGODB_QUIZ_QUESTIONS_COLLECTION_NAME="quiz_questions"
+# Note: We use the Azure-OpenAI style passthrough for the Radiant endpoint
+RADIANT_ENDPOINT="<base_url>/azure"
+RADIANT_API_KEY="rad-API_KEY-iant"
+# Cookie must be updated daily
+MONGODB_AUTH_COOKIE="<Auth cookie from CorpSecure-protected site>"
```
[the remark this answers is not in the document]
Comment (mongodben) [reply packages/model-eval/.env.example:11]: it's a manual process, but an easy one. for infosec purposes, i would prefer not to document it in our repo. i will send you a DM with the info

File: packages/model-eval/src/radiantModels.ts (hunk 2 above)
Comment (mongodben) [reply packages/model-eval/src/radiantModels.ts:25]: there's no specific ticket on our end for this issue, but i think it's generally captured by https://jira.mongodb.org/browse/EAI-379, which is to use more models on the evals

File: packages/mongodb-chatbot-evaluation/src/test/mockGenerateDataFunc.ts (hunk 3 above)
Comment (mongodben) [reply packages/mongodb-chatbot-evaluation/src/test/mockGenerateDataFunc.ts:30]: TIL about [Type Predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). i'd seen the `is` in but never really understood what was happening

File: packages/mongodb-chatbot-evaluation/src/generate/generateLlmQuizQuestionAnswer.ts (hunk 4 above)
Comment (mongodben) [reply packages/mongodb-chatbot-evaluation/src/generate/generateLlmQuizQuestionAnswer.ts:61]: HELM is just for quiz-style questions, the current `makeQuizQuestionPrompt()` does what you're describing. i can refactor comments a bit to make this clearer + also export this function and test it

File: packages/mongodb-chatbot-evaluation/src/test/mockGenerateDataFunc.ts (hunk 3 above)
Comment (mongodben) [reply packages/mongodb-chatbot-evaluation/src/test/mockGenerateDataFunc.ts:30]: this worked btw

**Pull request merged**